pkg/testing/runner/ssh.go

```diff
@@ -0,0 +1,207 @@
+}
+
+// newSSHPublicKey returns bytes for writing to .pub file
+func newSSHPublicKey(pk *rsa.PublicKey) ([]byte, error) {
+	pub, err := ssh.NewPublicKey(pk)
+	if err != nil {
+		return nil, err
+	}
+	return ssh.MarshalAuthorizedKey(pub), nil
+}
+
+// sshConnect keeps trying to make the SSH connection up until the context is cancelled
+func sshConnect(ctx context.Context, ip string, username string, sshAuth ssh.AuthMethod) (*ssh.Client, error) {
+	var lastErr error
+	for {
+		if ctx.Err() != nil {
+			if lastErr == nil {
+				return nil, ctx.Err()
+			}
+			return nil, lastErr
+		}
+		config := &ssh.ClientConfig{
+			User:            username,
+			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+			Auth:            []ssh.AuthMethod{sshAuth},
+			Timeout:         30 * time.Second,
+		}
+		client, err := ssh.Dial("tcp", net.JoinHostPort(ip, "22"), config)
+		if err == nil {
+			return client, nil
+		}
+		lastErr = err
+	}
+}
+
+// sshRunCommand runs a command on the SSH client connection
+func sshRunCommand(ctx context.Context, c *ssh.Client, cmd string, args []string, stdin io.Reader) ([]byte, []byte, error) {
+	if ctx.Err() != nil {
+		return nil, nil, ctx.Err()
+	}
+
+	cmdArgs := []string{cmd}
+	cmdArgs = append(cmdArgs, args...)
+	cmdStr := strings.Join(cmdArgs, " ")
+	session, err := c.NewSession()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer session.Close()
+
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	session.Stdout = &stdout
+	session.Stderr = &stderr
+	if stdin != nil {
+		session.Stdin = stdin
+	}
+	err = session.Run(cmdStr)
+	return stdout.Bytes(), stderr.Bytes(), err
+}
+
+// sshRunCommandWithRetry runs the command on loop waiting the interval between calls
+func sshRunCommandWithRetry(ctx context.Context, c *ssh.Client, cmd string, args []string, interval time.Duration) ([]byte, []byte, error) {
+	var lastErr error
+	var lastStdout []byte
+	var lastStderr []byte
+	for {
+		// the length of time for running the command is not blocked on the interval
+		// don't create a new context with the interval as its timeout
+		stdout, stderr, err := sshRunCommand(ctx, c, cmd, args, nil)
+		if err == nil {
+			return stdout, stderr, nil
+		}
+		lastErr = err
+		lastStdout = stdout
+		lastStderr = stderr
+
+		// wait for the interval or ctx to be cancelled
+		select {
+		case <-ctx.Done():
+			if lastErr != nil {
+				return lastStdout, lastStderr, lastErr
+			}
+			return nil, nil, ctx.Err()
+		case <-time.After(interval):
+		}
+	}
+}
+
+// sshSCP copies the filePath to the destination.
+func sshSCP(c *ssh.Client, filePath string, dest string) error {
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	s, err := f.Stat()
+	if err != nil {
+		return err
+	}
+
+	session, err := c.NewSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	w, err := session.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	cmd := fmt.Sprintf("scp -t %s", dest)
```

Comment:
Instead of copying the file ourselves, couldn't we just launch a command like
`scp <src file> <dst file>` 
where either the src or dst have the [scp://]user@host:/path format ?
Is it because we never write the ssh keys or config in a file to pass to the `scp` command ?

Reply:
Its so the executing host doesn't have to have scp and so we don't have to create yet another SSH connection when the runner already has one. Why use `scp` if we can easily do it here without shelling out to an external process? Its rather simple.